### PR TITLE
Updated to use source blocks

### PR DESCRIPTION
**Pull request by baweaver**
As per issue #5062

Noted that this isn't currently updating the specs and probably needs some work. Call this a prelim draft of what may need to be updated. I welcome any suggestions on improving it.

### CODE REVIEW

**Comment by segiddins** [comment]
This shouldn't be necessary if there's only a single RubyGems source in the Gemfile

File: lib/bundler/cli/install.rb [hunk 1]
```diff
@@ -100,7 +100,9 @@ def run
         Bundler.ui.warn <<-WARN, :wrap => true
           Your Gemfile has no gem server sources. If you need gems that are \
           not already on your machine, add a line like this to your Gemfile:
-          source 'https://rubygems.org'
+          source 'https://rubygems.org' do
```
Comment (cben) [inline lib/bundler/cli/install.rb]: Perhaps also reword above line from "add a line" to "wrap your Gemfile in a block like this" or something similar.

File: man/gemfile.5.ronn [hunk 2]
```diff
@@ -20,18 +20,19 @@ a number of methods used to describe the gem requirements.
...
-`source` lines. Each of these `source`s `MUST` be a valid Rubygems repository.
-
-Sources are checked for gems following the heuristics described in
-[SOURCE PRIORITY][]. If a gem is found in more than one global source, Bundler
-will print a warning after installing the gem indicating which source was used,
-and listing the other sources where the gem is available. A specific source can
-be selected for gems that need to use a non-standard repository, suppressing
-this warning, by using the [`:source` option](#SOURCE-source-) or a
-[`source` block](#BLOCK-FORM-OF-SOURCE-GIT-PATH-GROUP-and-PLATFORMS).
+Global source lines are a security risk and should not be used as they can lead
```
Comment (cben) [inline man/gemfile.5.ronn]: This doesn't mention single-gem `:source` options.
I think the audience the docs should be written for is not just for new people writing Gemfile from scratch, but people already having Gemfiles and trying to understand how they actually behave.

Worse, other places in the file do describe `:source` as something that sounds safe, eg.:

> This forces the gem to be loaded from this source and ignores any global sources

but [the blog post][1] specifically shows that:

```
source "http://public.org"
gem "foo"
gem "bar", source: "http://private.com"
gem "baz"
```
is unsafe.  I suppose all places mentioning `source` / `:source` should be touched.

Are `:source` exceptions safe if added inside a `source 'https://rubygems.org' do` block?  Are they always bad?

[1]: http://collectiveidea.com/blog/archives/2016/10/06/bundlers-multiple-source-security-vulnerability/

**Comment by deivid-rodriguez** [comment]
I agree with @segiddins that we don't need this. I think using a single global default gem source is fine?

@indirect Do you have any thoughts?

**Comment by indirect** [comment]
most Gemfiles only have one source, and don't need blocks. even if your Gemfile has more than one source, I think the generally ok way to go is one global source and other sources getting their own blocks. unless I'm missing something, anyway.

**Comment by cben** [comment]
This was all intended to educate how to use multiple sources safely without hitting the bundler 1.x security bug with global sources:

https://bundler.io/blog/2014/08/14/bundler-may-install-gems-from-a-different-source-than-expected-cve-2013-0334.html
https://collectiveidea.com/blog/archives/2016/10/06/bundlers-multiple-source-security-vulnerability

A lot of time has passed since.

- Did 2.x fully solve the issue?
- Are all 1.x affected?
- Have everybody migrated to 2.x?  (I think not yet?)

Current docs https://bundler.io/man/gemfile.5.html#SOURCE do mention the issue:

> Using the :source option for an individual gem will also make that source available as a possible global source for any other gems which do not specify explicit sources. Thus, when adding gems with explicit sources, it is recommended that you also ensure all other gems in the Gemfile are using explicit sources.

(is this still true in 2.x?!)

above quote *technically* disclaims the issue but not in sufficient detail, doesn't explain the 1.x/2.x history, doesnt define what exactly "explicit sources" means, doesn't educate *how* to write a mixed-source Gemfile that actually means what a human reader would think it means.

Are there separate URLs for 1.x and 2.x docs?

**Comment by deivid-rodriguez** [comment]
Yeah, a lot of time has passed indeed...

The definitions, as I understand them, are:

* Top level source: `source` without a block. For example, the omnipresent `source "https://rubygems.org"`.
* Explicit source: `source` with a block. For example, `source "http://my_gem_server" do gem "rails" end`.

I think the "Resolution" section is https://bundler.io/blog/2014/08/14/bundler-may-install-gems-from-a-different-source-than-expected-cve-2013-0334.html is not accurate and the Gemfile given as an example there is perfectly fine. The issue is when using multiple top level sources in the same Gemfile. When that happens, `bundler` will warn you, and will fail when bundler 3.0 is released.

**Comment by deivid-rodriguez** [comment]
Ok, I now read the other link you posted and I see where this all comes from. We need to study this and give a proper answer to your questions. I'll try to investigate it and get back here when I have more answers.

**Pull request closed without merging**